pysat/_instrument.py

```
@@ -381,72 +381,45 @@ def __setitem__(self, key, new):
-            elif hasattr(new, '__getitem__'):
-                # if isinstance(new, Series):
-                #     # series input, 1D variable
-                #     self.data[key] = new  
-                #     self.meta[key] = {}
-                if isinstance(new, DataFrame):
-                    # dataframe input allows input of multiple
-                    # 1D variables at once
-                    self.data[key] = new[key]
-                    for ke in key:
-                        self.meta[ke] = {}
-                elif isinstance(new[0], pds.DataFrame):
-                    # input is a list_like of frames
-                    # this is higher order data
-                    self.data[key] = new
-                    if key not in self.meta.keys_nD():
-                        # set up default metadata structure
-                        ho_meta = _meta.Meta()
-                        for ho_key in new[0]:
-                            ho_meta[ho_key] = {}
-                        self.meta[key] = ho_meta
-                else:
-                    # not one of special ones above
-                    # let pandas sort it out
-                    self.data[key] = new  
-                    self.meta[key] = {}
-            elif isinstance(key, str):
-                # didn't get a list_like thing of data
-                # but we were given a name
-                # try and add it
-                self.data[key] = new  
-                self.meta[key] = {}
-            else:
-                raise ValueError("No support for supplied input key")
+        # input dict must have data in 'data', 
+        # the rest of the keys are presumed to be metadata
+        in_data = new.pop('data')
+        if isinstance(in_data, pds.DataFrame):
+            # input is a dataframe
+            # multiple 1D datasets, no need for more
+            # provides a good filter for the next step of if chain
+            pass
+        elif isinstance(in_data[0], pds.DataFrame):
```

This will throw an exception if in_data is not iterable. If non-iterables are not allowed, you should throw an exception. If they are allowed, you should test for the presence of \_\_iter\_\_ on in_data before executing this line.